scripts/partition_manager.py

```diff
@@ -0,0 +1,328 @@
+            adr_map[k]['size'] = size_configs[k]
+
+
+def load_adr_map(adr_map, input_files, output_file_name, app_override_file):
+    for f in input_files:
+        img_conf = yaml.safe_load(f)
+        img_conf[list(img_conf.keys())[0]]['out_dir'] = path.dirname(f.name)
+        img_conf[list(img_conf.keys())[0]]['out_path'] = path.join(path.dirname(f.name), output_file_name)
+
+        adr_map.update(img_conf)
+    adr_map['app'] = dict()
+    adr_map['app']['placement'] = ''
+    adr_map['app']['out_dir'] = path.dirname(app_override_file)
+    adr_map['app']['out_path'] = app_override_file
+
+
+def set_addresses(reqs, solution, flash_size):
+    # First image starts at 0
+    reqs[solution[0]]['address'] = 0
+    for i in range(1, solution.index('app') + 1):
+        current = solution[i]
+        previous = solution[i - 1]
+        reqs[current]['address'] = reqs[previous]['address'] + reqs[previous]['size']
+
+    has_image_after_app = len(solution) > solution.index('app') + 1
+    if has_image_after_app:
+        reqs[solution[-1]]['address'] = flash_size - reqs[solution[-1]]['size']
+        for i in range(len(solution) - 2, solution.index('app'), -1):
+            current = solution[i]
+            previous = solution[i + 1]
+            reqs[current]['address'] = reqs[previous]['address'] - reqs[current]['size']
+        reqs['app']['size'] = reqs[solution[solution.index('app') + 1]]['address'] - reqs['app']['address']
+    else:
+        reqs['app']['size'] = flash_size - reqs['app']['address']
+
+
+def set_sub_partition_address_and_size(reqs, sub_partitions):
+    for sp_name, sp_values in sub_partitions.items():
+        size = reqs[sp_values['inside']]['size'] // len(sp_values['sub_partitions'])
+        address = reqs[sp_values['inside']]['address']
+        for sub_partition in sp_values['sub_partitions']:
+            sp_key_name = "%s_%s" % (sp_name, sub_partition)
+            reqs[sp_key_name] = dict()
+            reqs[sp_key_name]['size'] = size
+            reqs[sp_key_name]['address'] = address
+            address += size
+
+
+def write_override_files(adr_map):
+    for img, conf in adr_map.items():
+        if 'out_path' not in conf.keys() or 'out_dir' not in conf.keys():
+            continue  # The 'image' being inspected is not an executable, just a place-holder.
+        open(conf['out_path'], 'w').write('''\
+#undef CONFIG_FLASH_BASE_ADDRESS
+#define CONFIG_FLASH_BASE_ADDRESS %s
+#undef CONFIG_FLASH_LOAD_OFFSET
+#define CONFIG_FLASH_LOAD_OFFSET 0
+#undef CONFIG_FLASH_LOAD_SIZE
+#define CONFIG_FLASH_LOAD_SIZE %s
+''' % (hex(conf['address']), hex(conf['size'])))
+
+
+def get_flash_size(config):
+    config.seek(0)  # Ensure that we search the entire file
+    for line in config.readlines():
+        match = re.match(r'#define CONFIG_FLASH_SIZE (\d*)', line)
+        if match:
+            return int(match.group(1)) * 1024
+    raise RuntimeError("Unable to find 'CONFIG_FLASH_SIZE' in any of: %s" % config.name)
```

Comment:
@mbolivar : I'd like to use west's BuildConfiguration instead of what we are doing here.

How can we import BuildConfiguration from here?

Reply:
Instead of parsing the kconfig file?

IMO it should be kept like this because 1 - we are already parsing this file, 2 - there exists a set of people which prefer not to keep west up to date, or think about west at all, and I don't see any benefit of mixing west in to the partition manager here.

Reply:
> How can we import BuildConfiguration from here?

The `runners` package just needs to be on `sys.path`.

> IMO it should be kept like this because 1 - we are already parsing this file, 2 - there exists a set of people which prefer not to keep west up to date, or think about west at all, and I don't see any benefit of mixing west in to the partition manager here.

The `runners` package is used by the west extension commands provided with Zephyr, but it isn't part of west at all. It is automatically kept up to date because it's in the zephyr repository in `zephyr/scripts/west_commands/runners`. We could promote it to `scripts` to use it elsewhere at only a slight extra cost of having to manipulate `sys.path` to get at it in the west extensions.

So this doesn't require keeping west up to date, or for people to install west, or mixing west into the partition manager.

Reply:
Wait, no, I am totally lying: the runners package does import west packages, even though it's not part of west. Sorry; I keep forgetting that.

However, we could factor out the common bits from `runners.core` that other scripts in zephyr want to use, and I think that would be a good idea. Thoughts?

Reply:
> However, we could factor out the common bits from runners.core that other scripts in zephyr want to use, and I think that would be a good idea. Thoughts?

I agree. Sounds like this won't be ready for this PR though :(